tcpclient/connectionsManagement.go

```diff
@@ -21,38 +22,54 @@ func reportConnectionStatus(debugOut io.Writer, statusChannel chan<- Connection,
 // TCPConnect just opens a TCP connection against the target described by
 // the host:port, and considers the id to report back status changes through the
 // status goChannel with descriptors matching the Connection struct supplied in this
-// same package.
-// It also needs an iowriter to print debugging information.
-func TCPConnect(id int, host string, port int, wg *sync.WaitGroup, debugOut io.Writer, statusChannel chan<- Connection) error {
+// same package. It also needs an iowriter to print debugging information (it can be
+// ioutil.Discard.
+func TCPConnect(id int, host string, port int, wg *sync.WaitGroup, debugOut io.Writer,
+		statusChannel chan<- Connection, closeRequest <-chan bool) error {
 	connectionDescription := Connection{
 		Id:     id,
-		status: ConnectionDialing,
+		status: connectionDialing,
 	}
 	reportConnectionStatus(debugOut, statusChannel, connectionDescription)
 	conn, err := net.DialTimeout("tcp", host+":"+strconv.Itoa(port),
-		time.Duration(defaultDialTimeoutInSecs)*time.Second)
+		time.Duration(DefaultDialTimeoutInSecs)*time.Second)
 	if err != nil {
-		connectionDescription.status = ConnectionError
+		connectionDescription.status = connectionError
 		reportConnectionStatus(debugOut, statusChannel, connectionDescription)
 		fmt.Fprintln(debugOut, "Connection", id, "was unable to open the connection. Error:")
 		fmt.Fprintln(debugOut, err)
 		wg.Done()
 		return err
 	}
 
-	connectionDescription.status = ConnectionEstablished
+	defer conn.Close()
+	connectionDescription.status = connectionEstablished
 	reportConnectionStatus(debugOut, statusChannel, connectionDescription)
 	connBuf := bufio.NewReader(conn)
 	for {
-		str, err := connBuf.ReadString('\n')
-		if len(str) > 0 {
-			fmt.Fprintln(debugOut, "Connection", id, "got", str)
-		}
-		if err != nil {
-			connectionDescription.status = ConnectionClosed
-			reportConnectionStatus(debugOut, statusChannel, connectionDescription)
+		select {
+		case <-closeRequest:
+			fmt.Fprintln(debugOut, "Connection", id, "is being requested to close")
```

Comment:
when we get a closeRequest from the user, I understand that is not the `connection` that's being closed but the program itself, is this true?

Reply:
In this context, we are just opening/managing a single tcp connection, and yes, in this specific part of the logic what is being closed, not the whole program (MultiTCPConnect is, more or less, like the "main thread")